### PR TITLE
Person to card

### PR DESCRIPTION
**Pull request by ChangDarren**
Initiating  the change from Person object in AB4 to card object. The end goal is to change the list of Persons in AB4 to a list of cards. Deck structure will not be implemented in this PR. 

Not ready yet, but draft PR is called so that everyone can see and comment on how they think would be best to structure the card.

### CODE REVIEW

File: src/test/java/seedu/address/testutil/TypicalCards.java [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+import java.util.List;
+
+import seedu.address.model.deck.Card;
+
+/**
+ * A utility class containing a list of {@code Card} objects to be used in tests.
+ */
+public class TypicalCards {
+    public static final Card ADDITION = new CardBuilder().withQuestion("What is 1 + 1?")
+            .withAnswer("2").withTags("Easy").build();
```
Comment (xsot) [inline src/test/java/seedu/address/testutil/TypicalCards.java]: You should avoid using "easy" as a tag as difficulty will be a special type of tag in the future.

File: src/main/java/seedu/address/model/deck/Card.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+ * Represents a deck object in TopDeck.
+ * Guarantees: details are present and not null, field values are validated and immutable
+ */
+public class Card {
+    //Identity field
+    private final String question;
+
+    private final String answer;
+
+    private Set<Tag> tags = new HashSet<Tag>();
```
Comment (xsot) [inline src/main/java/seedu/address/model/deck/Card.java]: Why isn't this final?

File: src/main/java/seedu/address/model/deck/Card.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+        Card otherCard = (Card) other;
+        return this.answer.equals(otherCard.answer)
+                && this.question.equals(otherCard.question);
+    }
+
+    @Override
+    public int hashCode() {
+        int hash = 7;
+        return hash + question.hashCode() + answer.hashCode();
```
Comment (xsot) [inline src/main/java/seedu/address/model/deck/Card.java]: Better use the library implementation: `Objects.hash()`

File: src/main/java/seedu/address/model/deck/Card.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+        return otherCard != null
+                && otherCard.getQuestion().equals(question);
+    }
+
+
+    /**
+     * Returns true if both cards have the same question and answers.
+     * Defines a stronger notion of equality between 2 cards.
+     */
```
Comment (xsot) [inline src/main/java/seedu/address/model/deck/Card.java:67]: You forgot to check tags.

File: src/main/java/seedu/address/model/deck/Card.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+     * Returns an immutable tag set, which throws {@code UnsupportedOperationException}
+     * if modification is attempted.
+     */
+    public Set<Tag> getTags() {
+        return Collections.unmodifiableSet(tags);
+    }
+
+    /**
+     * Returns true if both cards have the same question.
+     * This defines a weaker notion of equality between two persons.
```
Comment (xsot) [inline src/main/java/seedu/address/model/deck/Card.java]: Remember to change all instances of 'person' to 'card'.

File: src/test/java/seedu/address/model/deck/CardTest.java [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+        Card card = new CardBuilder().build();
+        thrown.expect(UnsupportedOperationException.class);
+        card.getTags().remove(0);
+    }
+
+    @Test
+    public void isSameCard() {
+        Card additionCopy = new CardBuilder(ADDITION).build();
+
+        //same object -> returns true
```
Comment (xsot) [inline src/test/java/seedu/address/model/deck/CardTest.java]: Missing space after `//`

**Comment by ChangDarren** [comment]
I have changed the parts for the model and storage portion. The Logic portion of the code base however is still running base on AB4 and so the test cases will fail. Is it possible to push this as a seperate branch and then someone else help me to proceed with changing the logic section? Thank you.

**Comment by ChangDarren** [comment]
The checks that have not been updated are:
* logic/commands
* UI

**Comment by ChangDarren** [comment]
Things left:
1. Fix the System test case for Edit, Find and Delete.
2. Replace the last remnants of AB4 (persons) to TopDeck (Card)

The UI and app is working generally, so nothing huge is broken. Yay!!

**Comment by ChangDarren** [comment]
Do not merge the branch yet as there are still remnants of old code from AddressBook.

**Pull request merged**